File: 66.Dinic'sAlgo.java
/* ----------------------------------------------------------------------------  */
/*   ( The Authentic JS/JAVA CodeBuff )
 ___ _                      _              _ 
 | _ ) |_  __ _ _ _ __ _ __| |_ __ ____ _ (_)
 | _ \ ' \/ _` | '_/ _` / _` \ V  V / _` || |
 |___/_||_\__,_|_| \__,_\__,_|\_/\_/\__,_|/ |
                                        |__/ 
 */
/* --------------------------------------------------------------------------   */
/*    Youtube: https://youtube.com/@code-with-Bharadwaj                        */
/*    Github : https://github.com/Manu577228                                  */
/*    Portfolio : https://manu-bharadwaj-portfolio.vercel.app/portfolio       */
/* -----------------------------------------------------------------------  */

import java.io.*;
import java.util.*;

class Edge {
    int v;            // destination vertex
    int capacity;     // edge capacity
    int flow;         // current flow
    Edge rev;         // reverse edge

    Edge(int v, int capacity) {
        this.v = v;
        this.capacity = capacity;
        this.flow = 0;
    }
}

class Dinic {
    private int n;                   // number of vertices
    private List<Edge>[] graph;      // adjacency list

    @SuppressWarnings("unchecked")
    Dinic(int n) {
        this.n = n;
        graph = new ArrayList[n];
        for (int i = 0; i < n; i++) graph[i] = new ArrayList<>();
    }

    // Step 1: Add edge (and reverse edge for residual graph)
    void addEdge(int u, int v, int capacity) {
        Edge a = new Edge(v, capacity);
        Edge b = new Edge(u, 0);
        a.rev = b;
        b.rev = a;
        graph[u].add(a);
        graph[v].add(b);
    }

    // Step 2: Build level graph using BFS
    boolean bfs(int s, int t, int[] level) {
        Arrays.fill(level, -1);
        level[s] = 0;
        Queue<Integer> q = new LinkedList<>();
        q.add(s);
        while (!q.isEmpty()) {
            int u = q.poll();
            for (Edge e : graph[u]) {
                if (level[e.v] < 0 && e.flow < e.capacity) {
                    level[e.v] = level[u] + 1;
                    q.add(e.v);
                }
            }
        }
        return level[t] >= 0;
    }

    // Step 3: DFS to send blocking flow
    int dfs(int u, int t, int flow, int[] level, int[] start) {
        if (u == t) return flow;
        for (; start[u] < graph[u].size(); start[u]++) {
            Edge e = graph[u].get(start[u]);
            if (level[e.v] == level[u] + 1 && e.flow < e.capacity) {
                int currFlow = Math.min(flow, e.capacity - e.flow);
                int pushed = dfs(e.v, t, currFlow, level, start);
                if (pushed > 0) {
                    e.flow += pushed;
                    e.rev.flow -= pushed;
                    return pushed;
                }
            }
        }
        return 0;
    }

    // Step 4: Main max flow function
    int maxFlow(int s, int t) {
        if (s == t) return 0;
        int total = 0;
        int[] level = new int[n];

        while (bfs(s, t, level)) {
            int[] start = new int[n];
            while (true) {
                int flow = dfs(s, t, Integer.MAX_VALUE, level, start);
                if (flow == 0) break;
                total += flow;
            }
        }
        return total;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        Dinic dinic = new Dinic(6);

        dinic.addEdge(0, 1, 16);
        dinic.addEdge(0, 2, 13);
        dinic.addEdge(1, 2, 10);
        dinic.addEdge(2, 1, 4);
        dinic.addEdge(1, 3, 12);
        dinic.addEdge(3, 2, 9);
        dinic.addEdge(2, 4, 14);
        dinic.addEdge(4, 3, 7);
        dinic.addEdge(3, 5, 20);
        dinic.addEdge(4, 5, 4);

        System.out.println("Maximum Flow: " + dinic.maxFlow(0, 5));
    }
}

File: 66.Dinic'sAlgo.py
# Dinic’s Algorithm is an efficient method to compute the maximum flow in a flow network.
# It builds a level graph using BFS and then sends blocking flows using DFS until 
# no more augmenting paths exist.

# Explanation

# Imagine we have a network of pipes (edges) connecting tanks (nodes), 
# each pipe having a capacity (maximum flow it can carry).
# Our goal is to push as much flow as possible from a source (s) to a sink (t) 
# without exceeding the pipe capacities.

# Dinic’s Algorithm works in two phases repeatedly:

# Level Graph Construction (BFS) → find shortest distances (in edges) from s to all reachable nodes.

# Blocking Flow Computation (DFS) → send flow only along those shortest paths until we can’t send anymore.

# This process repeats until no augmenting paths remain, giving us the maximum flow.

from collections import deque

class Edge:
    def __init__(self, v, capacity):
        self.v = v
        self.capacity = capacity
        self.flow = 0
        self.rev = None

class Dinic:
    def __init__(self, n):
        self.n = n
        self.graph = [[] for _ in range(n)]

    def add_edge(self, u, v, capacity):
        forward = Edge(v, capacity)
        backward = Edge(u, 0)
        forward.rev = backward
        backward.rev = forward
        self.graph[u].append(forward)
        self.graph[v].append(backward)

    def bfs(self, s, t, level):
        queue = deque([s])
        level[s] = 0
        while queue:
            u = queue.popleft()
            for e in self.graph[u]:
                if level[e.v] < 0 and e.flow < e.capacity:
                    level[e.v] = level[u] + 1
                    queue.append(e.v)
        return level[t] >= 0
    
    def dfs(self, u, t, flow, level, start):
        if u == t:
            return flow
        while start[u] < len(self.graph[u]):
            e = self.graph[u][start[u]]
            if level[e.v] == level[u] + 1 and e.flow < e.capacity:
                curr_flow = min(flow, e.capacity - e.flow)
                pushed = self.dfs(e.v, t, curr_flow, level, start)
                if pushed > 0:
                    e.flow += pushed
                    e.rev.flow -= pushed
                    return pushed
            start[u] += 1
        return 0
    
    def max_flow(self, s, t):
        total = 0
        while True:
            level = [-1] * self.n
            if not self.bfs(s, t, level):
                break
            start = [0] * self.n
            while True:
                flow = self.dfs(s, t, float('inf'), level, start)
                if flow == 0:
                    break
                total += flow
        return total
    
if __name__ == "__main__":
    n = 6
    dinic = Dinic(n)
    dinic.add_edge(0, 1, 16)
    dinic.add_edge(0, 2, 13)
    dinic.add_edge(1, 2, 10)
    dinic.add_edge(2, 1, 4)
    dinic.add_edge(1, 3, 12)
    dinic.add_edge(3, 2, 9)
    dinic.add_edge(2, 4, 14)
    dinic.add_edge(4, 3, 7)
    dinic.add_edge(3, 5, 20)
    dinic.add_edge(4, 5, 4)

    print("Maximum Flow:", dinic.max_flow(0, 5))
        

        